layers/+completion/auto-completion/packages.el

```diff
@@ -149,6 +151,67 @@
        (unless (eq auto-completion-enable-help-tooltip 'manual)
          (company-quickhelp-mode))))))
 
+(defun auto-completion/init-company-box ()
+  (use-package company-box
+    :hook '(company-mode . company-box-mode)
+    :commands 'company-box-doc-manually
```

Comment:
There might be a loading order issue in Spacemacs, because without the line below, when `.spacemacs` is opened, then the messages buffer shows:
>Eager macro-expansion failure: (void-function all-the-icons-faicon)

```suggestion
    :commands (all-the-icons-faicon company-box-doc-manually)
```
However it doesn't seem to be a good solution. The message now says:
>Warning: Eager macro-expansion skipped due to cycle:
  … => (load "company-box-icons.el") => (macroexpand-all …) => (macroexpand (eval-when-compile …)) => (load "company-box-icons.el")

Reply:
adding all-the-icons in the required packages seems solve this problem.

Reply:
I also got the warning to disappear by adding the following to the `auto-completion-packages` list:
```elisp
        (all-the-icons :toggle auto-completion-use-company-box)
```

It also required that a `auto-completion/init-all-the-icons` function is defined and it needs to call `(use-package all-the-icons)`.

I don't know if this is better, but I moved the `all-the-icons` related code from the 
`auto-completion/init-company-box` function to the `auto-completion/init-all-the-icons`  function.
And set it to load `:after company`.
That made it possible to remove:
- The `with-eval-after-load` line.
- And the `all-the-icons-faicon` entry from the `auto-completion/init-company-box` functions `:commands` section.

The `auto-completion/init-all-the-icons` function ended up looking like this:
```elisp
(defun auto-completion/init-all-the-icons ()
  (use-package all-the-icons
    :after company
    :config
    (progn
      (eval-and-compile
        (defun my-company-box-icon (family icon &rest args)
          "Defines icons using `all-the-icons' for `company-box'."
          (when icon
            (let ((icon (pcase family
                          ('octicon (all-the-icons-octicon icon :height 0.8 :v-adjust -0.05 args))
                          ('faicon (all-the-icons-faicon icon :height 0.8 :v-adjust -0.0575))
                          ('material (all-the-icons-material icon :height 0.8 :v-adjust -0.225 args))
                          ('alltheicon (all-the-icons-alltheicon icon :height 0.8 args)))))
              (unless (symbolp icon)
                (concat icon
                        (propertize " " 'face 'variable-pitch)))))))
      (setq company-box-icons-all-the-icons
            `((Unknown . ,(my-company-box-icon 'octicon "file-text"))
              (Text . ,(my-company-box-icon 'faicon "file-text-o"))
              (Method . ,(my-company-box-icon 'faicon "cube"))
              (Function . ,(my-company-box-icon 'faicon "cube"))
              (Constructor . ,(my-company-box-icon 'faicon "cube"))
              (Field . ,(my-company-box-icon 'faicon "tag"))
              (Variable . ,(my-company-box-icon 'faicon "tag"))
              (Class . ,(my-company-box-icon 'faicon "cog"))
              (Interface . ,(my-company-box-icon 'faicon "cogs"))
              (Module . ,(my-company-box-icon 'alltheicon "less"))
              (Property . ,(my-company-box-icon 'faicon "wrench"))
              (Unit . ,(my-company-box-icon 'faicon "tag"))
              (Value . ,(my-company-box-icon 'faicon "tag"))
              (Enum . ,(my-company-box-icon 'faicon "file-text-o"))
              (Keyword . ,(my-company-box-icon 'material "format_align_center"))
              (Snippet . ,(my-company-box-icon 'material "content_paste"))
              (Color . ,(my-company-box-icon 'material "palette"))
              (File . ,(my-company-box-icon 'faicon "file"))
              (Reference . ,(my-company-box-icon 'faicon "tag"))
              (Folder . ,(my-company-box-icon 'faicon "folder"))
              (EnumMember . ,(my-company-box-icon 'faicon "tag"))
              (Constant . ,(my-company-box-icon 'faicon "tag"))
              (Struct . ,(my-company-box-icon 'faicon "cog"))
              (Event . ,(my-company-box-icon 'faicon "bolt"))
              (Operator . ,(my-company-box-icon 'faicon "tag"))
              (TypeParameter . ,(my-company-box-icon 'faicon "cog"))
              (Template . ,(my-company-box-icon 'octicon "file-code")))))))
```

And the `auto-completion/init-company-box` function now looks like this:
- With the hide-lighter (mode line text) line added,
- It's missing the liguangsheng source url.
- I added a `progn` section under `:config` (that seems to be the convention when a `:init` or `:config` section has more than one line.)
```elisp
(defun auto-completion/init-company-box ()
  (use-package company-box
    :hook '(company-mode . company-box-mode)
    :commands 'company-box-doc-manually
    :init (setq company-box-icons-alist 'company-box-icons-all-the-icons)
    :config
    (progn
      (spacemacs|hide-lighter company-box-mode)
      ;; copied from @liguangsheng
      (setq company-box-backends-colors nil
            company-box--max 1000)

      (if (equal auto-completion-enable-help-tooltip nil)
          (setq company-box-doc-enable nil)
        (if (equal auto-completion-enable-help-tooltip 'manual)
            (let ()
              (setq company-box-doc-enable nil)
              (define-key company-active-map (kbd "M-h") #'company-box-doc-manually)))))))
```

The `if` statements seems like they could be simplified.

One possible change could be to change the second if statement to a: `when`
because it doesn't have an `else` section.
And the empty `(let ()`, isn't needed. `when` wraps the commands in a `progn`.
With these changes the function would look like this:
```elisp
(defun auto-completion/init-company-box ()
  (use-package company-box
    :hook '(company-mode . company-box-mode)
    :commands 'company-box-doc-manually
    :init (setq company-box-icons-alist 'company-box-icons-all-the-icons)
    :config
    (progn
      (spacemacs|hide-lighter company-box-mode)
      ;; copied from @liguangsheng
      (setq company-box-backends-colors nil
            company-box--max 1000)
      (if (equal auto-completion-enable-help-tooltip nil)
          (setq company-box-doc-enable nil)
        (when (equal auto-completion-enable-help-tooltip 'manual)
          (setq company-box-doc-enable nil)
          (define-key company-active-map (kbd "M-h") #'company-box-doc-manually))))))
```

Another possible change could be to:
Instead of setting the `company-box-doc-enable` variable to `nil` in two cases, it could be set to `nil` by default, and only enabled if `auto-completion-enable-help-tooltip` is set to `t`.
With these changes the function could look like this:
```elisp
(defun auto-completion/init-company-box ()
  (use-package company-box
    :hook '(company-mode . company-box-mode)
    :commands 'company-box-doc-manually
    :init (setq company-box-icons-alist 'company-box-icons-all-the-icons)
    :config
    (progn
      (spacemacs|hide-lighter company-box-mode)
      ;; copied from @liguangsheng
      (setq company-box-backends-colors nil
            company-box--max 1000
            company-box-doc-enable nil)
      (case auto-completion-enable-help-tooltip
        ('manual (define-key company-active-map
                   (kbd "M-h") #'company-box-doc-manually))
        ('t (setq company-box-doc-enable t))))))
```

While testing the second suggestion, I noticed that the `company-box--max` variable is set to `1000` before a company popup has been opened. But it's set to `47` afterwards.
I don't know where it gets set to `47`.

Reply:
Wow, thanks for the advice. My code is truly ill structured. I'm sorry that I don't have much experience writing elisp.
I think put icons settings in `auto-completion/init-all-the-icons` is a good idea because separating face settings and behavior settings seems more clear. since only `company-box` is using all-the-icons in the layer and all-the-icons doesn't require configuration.